Makefile and other visual changes.

kishen-v opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide/pull-requests.md#the-pull-request-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
5. If this PR changes image versions, please title this PR "Bump <image name> from x.x.x to y.y.y."

-->

**What this PR does / why we need it**: 
Changes in this PR:
1. `make serve` is now made available from the parent Makefile and can be executed from the repository root.
2. `README.md` changes to fix the broken link of License, and other related badges, alignment of the images, etc.
3.   Version bump of `mdbook` from v0.4.5 to v0.4.35.
4.  Add `make update-go` to update golang version across multiple files.
5.  Add `make verify-go-version` to find if there are any recent release of Go that can be updated.
```
Kishens-MacBook-Pro: cluster-api-provider-ibmcloud : mdbook-target ᐅ  make update-go VERSION=1.21.4 
-FROM --platform=${BUILDPLATFORM} golang:1.20.11 as toolchain
+FROM --platform=${BUILDPLATFORM} golang:1.21.4 as toolchain

-GO_VERSION ?= 1.20.11
+GO_VERSION ?= 1.21.4

-FROM golang:1.20.11 AS ccm-builder
+FROM golang:1.21.4 AS ccm-builder
```
6.  Update favicon image to show clusterapi logo, from the default mdbook logo.
<img width="549" alt="Pasted Graphic 4" src="https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/assets/110517346/4700b12b-fef0-4307-bc90-0252323bd2a9">


**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1510


**Special notes for your reviewer**: This PR has multiple commits. Please let me know if it needs to be made into separate PRs.

/area provider/ibmcloud


**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
3. If no release note is required, just write "NONE".
-->
```release-note
make serve is now made available from the parent Makefile and can be executed from the repository root.
README.md changes to fix the broken link of License, and other related badges, alignment of the images, etc.
Version bump of mdbook from v0.4.5 to v0.4.37.
Add make update-go to update golang version across Makefiles
Add make verify-go-version to find if inconsistencies exist across files with regards to go version in Make/Submakefiles.
Update favicon image to ClusterAPI logo.
```

kishen-v commented:
/cc @Amulyam24  @Prajyot-Parab.

mkumatag commented on Makefile:
```diff
@@ -600,3 +600,21 @@ kind-cluster: ## Create a new kind cluster designed for development with Tilt
+## --------------------------------------
+
+.PHONY: update-go
+update-go: ## Update Go version across files - Usage make update-version VERSION=X.YY.ZZ
```
can we somehow call this in the verify flow and ensure files are always updated?

kishen-v replied:
Hi @mkumatag, 
I had discussed with @Prajyot-Parab regarding this and have added a new target `verify-go-version` which checks if a patch release is available for the n-1'th minor release (Just the way we follow) and chose not to block PRs when the version check fails.

I have a single commit for this action in a different branch to trigger for all PR, which confirms if we're using the latest patch release.  I shall enable this action for this repository if this seems ok.
The current behavior is that a warning is raised in the workflow regarding the version mismatch.
<img width="937" alt="image" src="https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/assets/110517346/f6acb6a9-0526-49dd-b066-85ba7c77b344">

along with the following output in the shell:
```
go1.20.12 is available on go.dev, the version in Makefile is 1.20.11.
Consider bumping the version, run make update-go VERSION=go1.20.12 to update across all files.
```

The above check is a part of `verify` as well. Since we do not mark the job as failed, the findings related to version remains as a few lines of log in the prow-job.

mkumatag replied:
I feel allowing user to decide what version they want to use, we can update the version manually or by some automation like dependatbot will be better idea.

kishen-v commented on Makefile:
```diff
@@ -600,3 +600,21 @@ kind-cluster: ## Create a new kind cluster designed for development with Tilt
+## --------------------------------------
+
+.PHONY: update-go
+update-go: ## Update Go version across files - Usage make update-version VERSION=X.YY.ZZ
```
Really considering `make update-version X.Y.Z` than `make update-version VERSION=X.YY.ZZ`.
Will look into it.

Prajyot-Parab replied:
@kishen-v any update?

Prajyot-Parab commented on versions.mk:
```diff
@@ -12,4 +12,4 @@
 # limitations under the License.
 
-MDBOOK_VERSION := v0.4.5
+MDBOOK_VERSION := v0.4.35
```
```suggestion
MDBOOK_VERSION := v0.4.36
```
New release is out!

kishen-v commented on Makefile:
```diff
@@ -25,6 +25,8 @@ IMG ?= controller:latest
 # Produce CRDs that work back to Kubernetes 1.11 (no version conversion)
 CRD_OPTIONS ?= "crd:crdVersions=v1"
 
+SHELL=/bin/bash
```
Observing `/bin/sh: 1: Syntax error: redirection unexpected` as here-strings are used in verify-go-version.
 Hence, setting shell to bash shell from user shell as a workaround.

kishen-v commented on Makefile:
```diff
@@ -539,6 +541,26 @@ verify-security: ## Verify code and images for vulnerabilities
+	IFS=' ' read -r current_minor current_patch <<< "$$current_minor_patch_version"; \
+	if [ $$current_minor -le $$available_minor ]; then \
+		if [ $$current_patch -lt $$available_patch ] || [ $$current_minor -lt $$available_minor ]; then \
+			echo "::warning ::Update GO_VERSION to $$available_version"; \
```
Ref: https://stackoverflow.com/a/75849954
<img width="920" alt="image" src="https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/assets/110517346/9b0f0462-8207-4e5e-b8b8-5dfa714e079e">

Prajyot-Parab commented:
@kishen-v 
Is this the expected behaviour?

```
prajyotparab@Prajyots-MacBook-Pro cluster-api-provider-ibmcloud % make verify-go-version
fatal: No names found, cannot describe anything.
::warning ::Update GO_VERSION to go1.20.13
go1.20.13 is available on go.dev, the version in Makefile is 1.20.12.
Consider bumping the version, run make update-go VERSION=go1.20.13 to update across all files.
```

Karthik-K-N reviewed (commented):
/lgtm

@Prajyot-Parab Please take a look

kishen-v commented:
@Prajyot-Parab, If it's this line `fatal: No names found, cannot describe anything.` 
I see the same error on my machine. I shall look into it.

Prajyot-Parab commented:
> @Prajyot-Parab, If it's this line `fatal: No names found, cannot describe anything.` I see the same error on my machine. I shall look into it.

Yes, please check.

kishen-v commented:
@Prajyot-Parab, I see it now.. git describe seems to be causing a problem. 
```
Kishens-MacBook-Pro:  cluster-api-provider-ibmcloud : mdbook-target: git describe
fatal: No names found, cannot describe anything.
```
added `--always`  to pick the latest commit object as fallback to supress the error.
```
Kishens-MacBook-Pro: cluster-api-provider-ibmcloud : mdbook-target :  git describe --always --abbrev=0
44b176a7ffd2a0253429e29c53da1613924bbca2
```

Doc:  https://git-scm.com/docs/git-describe#Documentation/git-describe.txt---always

With the above change in the Makefile and version.sh, the error is no longer seen. 
```
Kishens-MacBook-Pro:  cluster-api-provider-ibmcloud : mdbook-target:  make verify-go-version
::warning ::Update GO_VERSION to go1.20.13
go1.20.13 is available on go.dev, the version in Makefile is 1.20.12.
Consider bumping the version, run make update-go VERSION=go1.20.13 to update across all files.
```

kishen-v commented:
@Prajyot-Parab, ptal! Required changes have been made and pushed to branch.

Prajyot-Parab commented on Makefile:
Need to fix:

```
::warning ::Update GO_VERSION to go1.20.13
go1.20.13 is available on go.dev, the version in Makefile is 1.20.12.
Consider bumping the version, run make update-go VERSION=go1.20.13 to update across all files.
prajyotparab@Prajyots-MacBook-Pro cluster-api-provider-ibmcloud % make update-go VERSION=go1.20.13
```

it should suggest -- > ```make update-go VERSION=1.20.13```

kishen-v replied:
Thanks for pointing out! 
The changes are made to print just the version.
```
Kishens-MacBook-Pro:  cluster-api-provider-ibmcloud : mdbook-target :  make verify-go-version
::warning ::Update GO_VERSION to 1.20.13
1.20.13 is available on go.dev, the version in Makefile is 1.20.12.
Consider bumping the version, run "make update-go VERSION=1.20.13" to update across all files.
```

Prajyot-Parab commented on Makefile:
```diff
@@ -18,13 +18,15 @@ ROOT_DIR_RELATIVE := .
 
 include $(ROOT_DIR_RELATIVE)/common.mk
 
-GO_VERSION ?= 1.20.12
```
@kishen-v why this change?

kishen-v replied:
Sorry, it was an unintended change which went along with latest commit. 
I have reverted back to the version on `main`.

Prajyot-Parab reviewed (approved):
/lgtm

@Karthik-K-N ptal

Karthik-K-N reviewed (approved):
Thanks for the PR

/lgtm

kishen-v commented:
Hi @Prajyot-Parab/ @Karthik-K-N,

I have incorporated the discussed changes where:
1. Do away with go version set in the Makefile.
2. Retrieve the go-version from the Dockerfile in repository root.
3. Update go-version across all Dockerfiles in the repository.

Here are the o/ps.
```
Updating versions across all dockerfiles.

Kishens-MacBook-Pro:  cluster-api-provider-ibmcloud : mdbook-target :  make update-go VERSION=1.20.12
Updated go version to 1.20.12 across Dockerfile hack/ccm/Dockerfile

Verify that Go versions are same across all Dockerfiles

Kishens-MacBook-Pro:  cluster-api-provider-ibmcloud : mdbook-target : make verify-go-version
Versions are uniform across Dockerfiles, the go-version used is 1.20.12

And if not...

Kishens-MacBook-Pro: cluster-api-provider-ibmcloud : mdbook-target  :  make verify-go-version
Versions are different across Dockerfiles, please use 'make update-go'
```

cc: @mkumatag.

mkumatag reviewed (approved):
I will let @Prajyot-Parab to add lgtm

Prajyot-Parab commented:
@kishen-v lets do it this way
1. Refer https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/Makefile#L23-L24 ,lets use same approach here, we will pass the image itself as build-arg to our docker commands and that way no need to fetch any version from Dockerfiles.
2. Lets have check in place to verify the version mentioned in Makefile and ccm dockerfile is same, If yes all good, If no just throw out warning saying version mismatch and suggest user to have consistency.
Also you can refer this older commit (https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/commit/3e0af2a737146aff17a2b268ecd285dd8e74c992) lets switch to this in ccm Dockerfile (this will avoid triggering update via depenbot), you can also pass that image as arg via ccm Makefile and have comparison done on it rather than Dockerfile
3. Lets not have any command in place to update the versions via make  (drop that logic)

let me know incase of any queries.

Prajyot-Parab commented:
@kishen-v squash everything into 1 commit. (rest lgtm)

Prajyot-Parab commented:
/hold

Prajyot-Parab reviewed (approved):
/lgtm

Prajyot-Parab commented:
/unhold
